metadata/src/retain.rs

```diff
@@ -0,0 +1,209 @@
+    let old_id = ty.id();
+    let new_id = map_ids.get(&old_id).expect(&format!(
+        "PortableRegistry did not retain type id {old_id}. This is a bug. Please open an issue."
+    ));
+    *ty = (*new_id).into();
+}
+
+/// Update all type IDs of the provided pallet using the new type IDs from the portable registry.
+fn update_pallet_types(pallet: &mut PalletMetadata<PortableForm>, map_ids: &BTreeMap<u32, u32>) {
+    if let Some(storage) = &mut pallet.storage {
+        for entry in &mut storage.entries {
+            match &mut entry.ty {
+                StorageEntryType::Plain(ty) => {
+                    update_type(ty, map_ids);
+                }
+                StorageEntryType::Map { key, value, .. } => {
+                    update_type(key, map_ids);
+                    update_type(value, map_ids);
+                }
+            }
+        }
+    }
+
+    if let Some(calls) = &mut pallet.calls {
+        update_type(&mut calls.ty, map_ids);
+    }
+
+    if let Some(event) = &mut pallet.event {
+        update_type(&mut event.ty, map_ids);
+    }
+
+    for constant in &mut pallet.constants {
+        update_type(&mut constant.ty, map_ids);
+    }
+
+    if let Some(error) = &mut pallet.error {
+        update_type(&mut error.ty, map_ids);
+    }
+}
+
+/// Update all type IDs of the provided extrinsic metadata using the new type IDs from the portable registry.
+fn update_extrinsic_types(
+    extrinsic: &mut ExtrinsicMetadata<PortableForm>,
+    map_ids: &BTreeMap<u32, u32>,
+) {
+    update_type(&mut extrinsic.ty, map_ids);
+
+    for signed in &mut extrinsic.signed_extensions {
+        update_type(&mut signed.ty, map_ids);
+        update_type(&mut signed.additional_signed, map_ids);
+    }
+}
+
+/// Generate a subset of the metadata that contains only the
+/// types needed to represent the provided pallets.
+///
+/// # Note
+///
+/// Used to strip metadata of unneeded information and to reduce the
+/// binary size.
+///
+/// # Panics
+///
+/// Panics if the [`scale_info::PortableRegistry`] did not retain all needed types.
+pub fn retain_metadata_pallets<F>(metadata: &mut RuntimeMetadataV14, mut filter: F)
+where
+    F: FnMut(&PalletMetadata<PortableForm>) -> bool,
+{
+    let mut type_ids = HashSet::new();
+
+    // Collect all types needed to represent the provided pallets.
+    metadata.pallets.retain(|pallet| {
+        if filter(pallet) {
+            collect_pallet_types(pallet, &mut type_ids);
+            true
+        } else {
+            false
+        }
+    });
+
+    collect_extrinsic_types(&metadata.extrinsic, &mut type_ids);
+    type_ids.insert(metadata.ty.id());
+
+    // Additionally, subxt depends on the `RuntimeError` and `DispatchError`.
+    for ty in metadata.types.types() {
+        let Some(ident) = ty.ty().path().ident() else {
+            continue
+        };
+        if ident == DISPATCH_ERROR || ident == RUNTIME_EVENT {
+            type_ids.insert(ty.id());
+        }
```

Comment:
This also makes me wonder; are there any other types we might realise we need that aren't "connected" to the metadata in any way? How are they generated I wonder; do they happen to be used in certain calls?

As an earlier PR I pointed to the existing RuntimeEvent rather than creating a new one, but if the RuntimeEvent was only there by chance because it happened to be used in some pallet thing, that's probably a bad idea..

Reply:
Ah ok so `RuntimeEvent` is generates because it's returned from `system::storage::events` normally.

I think we should go back to generating a `RuntimeEvent` rather than pointing to that one then in Subxt codegen; then when we strip pallets we can generate only the event variants we are keeping, and we don't need to keep hold of the "RuntimeEvent" type unless it's actually used elsewhere in metadata.

Re "DispatchError", we do need that type because we use it to help decode our errors. We find it in metadata with:

```
        let dispatch_error_ty = metadata
            .types
            .types()
            .iter()
            .find(|ty| ty.ty().path().segments() == ["sp_runtime", "DispatchError"])
            .map(|ty| ty.id());
```

So perhaps we should do the same search here to be accurate?